test/spmd/tensor/parallel/test_tp_examples.py

```diff
@@ -371,5 +364,164 @@ def test_self_attn_megatron_e2e(self):
         self.assertEqual(output, output_tp)
 
 
+    # TensorParallelMultiheadAttention == dist_module(torch.nn.MultiheadAttention)
```

Comment:
Ah that's a good question. Your understanding is correct, `MultiheadAttnWrap` is simply a wrapper over `torch.nn.MultiheadAttention`. In short, we wrap `MultiheadAttention` because `_shard_self_attn` cannot replace the top level module in the module tree with our custom `TensorParallelMultiheadAttention` module.

In detail, if you look into the `_shard_self_attn` (https://github.com/pytorch/tau/blob/self-attn/spmd/tensor/parallel/api.py#L26) which we use to parallelize each `TensorParallelMultiheadAttention`/`torch.nn.MultiheadAttention` module in tree, you'll see that we search the immediate child nodes (`module.named_children()`) to find out those target modules, replace with `TensorParallelMultiheadAttention` if the module is a `torch.nn.MultiheadAttention`, and parallelize it in case of a `TensorParallelMultiheadAttention` module. The problem happens when we need to replace a module. What we currently adopt is, we register the new module under the same name within its parent module, and this determines that we cannot replace the top-level module whose parent module doesn't exist. However, unlikely will people use this API to parallelize a sole `torch.nn.MultiheadAttention` (simply not its use case) so we can just wrap it in our test. In real world cases where `torch.nn.MultiheadAttention` is used as a component, the wrap is not needed.

Reply:
If so, let's call it `DummyModule` or let's create a dummy transformer here. one MultiheadAttn + one MLP. WDYT?